java-frontend/src/main/java/org/sonar/java/JavaFrontend.java

```diff
@@ -0,0 +1,643 @@
+
+  /**
+   * Literals
+   * Non-string operands
+   */
+  public static class TaintFreeSource implements TaintSource {
+
+    @Override
+    public boolean canBeTainted() {
+      return false;
+    }
+
+    @Override
+    public TaintSource unionWith(TaintSource other) {
+      return other;
+    }
+
+    @Override
+    public String toString() {
+      return "taint-free";
+    }
+
+    @Override
+    public int hashCode() {
+      return 0;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (obj == null || !(obj.getClass().equals(this.getClass()))) {
+        return false;
+      }
+
+      return true;
+    }
+
+  }
+
+  /**
+   * Read-from method parameter
+   * Read-from field
+   * Read-from method-call return value
+   */
+  public static class DirectTaintSource implements TaintSource {
+
+    private final String signature;
+    private final List<TaintSource> arguments;
+
+    public DirectTaintSource(String signature, List<TaintSource> arguments) {
+      this.signature = signature;
+      this.arguments = arguments;
+    }
+
+    @Override
+    public boolean canBeTainted() {
+      return true;
+    }
+
+    @Override
+    public TaintSource unionWith(TaintSource other) {
+      return IndirectTaintSource.union(this, other);
+    }
+
+    @Override
+    public String toString() {
+      return signature + (arguments.isEmpty() ? "" : "(" + Joiner.on(", ").join(arguments) + ")");
+    }
+
+    @Override
+    public int hashCode() {
+      return signature.hashCode();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (obj == null || !(obj.getClass().equals(this.getClass()))) {
+        return false;
+      }
+
+      DirectTaintSource other = (DirectTaintSource) obj;
+
+      return signature.equals(other.signature) &&
+        arguments.equals(other.arguments);
+    }
+
+  }
+
+  /**
+   * String concatenation
+   */
+  public static class IndirectTaintSource implements TaintSource {
+
+    private final TaintSource left;
+    private final TaintSource right;
+
+    private IndirectTaintSource(TaintSource left, TaintSource right) {
+      this.left = left;
+      this.right = right;
+    }
+
+    @Override
+    public boolean canBeTainted() {
+      return true;
+    }
+
+    @Override
+    public TaintSource unionWith(TaintSource other) {
+      return union(this, other);
+    }
+
+    public static TaintSource union(TaintSource left, TaintSource right) {
+      if (!left.canBeTainted()) {
+        return right;
+      }
+
+      if (!right.canBeTainted()) {
+        return left;
+      }
+
+      if (left.equals(right)) {
+        return left;
+      }
+
+      return new IndirectTaintSource(left, right);
+    }
+
+    @Override
+    public String toString() {
+      return "left: " + left.toString() + ", right = " + right.toString();
+    }
+
+    @Override
+    public int hashCode() {
+      return left.hashCode() + right.hashCode();
+    }
+
+    // TODO equals is not symmetrical
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)